# Logic-Solver

# ⚠️ Setup Requirement: Type Definitions

To run this notebook with strict type checking, TypeScript needs type definitions for the `logic-solver` library, which are not included by default.

**Instructions:**
1. Run the code cell below to find the installation path of the `logic-solver` package on your system.
2. Locate the file `logic-solver.d.ts` provided in the current directory.
3. Move or copy this `logic-solver.d.ts` file into the directory path output by the cell below (inside the `logic-solver` folder).
4. Restart the kernel

Once the declaration file is in place, the notebook will compile and run correctly.

In [1]:
require.resolve('logic-solver');

/Users/karl/Library/Mobile Documents/com~apple~CloudDocs/Box/node_modules/logic-solver/logic-solver.js


[Logic-Solver](https://github.com/meteor/logic-solver) is a SAT solver available as an [npm package](https://www.npmjs.com/package/logic-solver). It is a JavaScript wrapper that relies on MiniSat designed for logical constraint satisfaction. We can install it via `npm install logic-solver`.

In [2]:
import * as Logic from 'logic-solver';

In this notebook, the [logic-solver](https://github.com/meteor/logic-solver) library is used to efficiently solve propositional logic (SAT) problems.

- **Boolean variables** are implicitly declared by using strings (e.g., `"p"`, `"q"`, `"r"`).
- **Clauses** are constructed using `Logic.or(...)`, with strings serving as literals.
- **Negations** are represented by a preceding minus sign in the string (e.g., `"-q"` for $\neg q$). Alternatively, `Logic.not(...)` could be used.
- The problem is built as a set of clauses (Conjunctive Normal Form, CNF) by adding each clause individually using `solver.require(...)`. This corresponds to a large AND over all clauses.
- The solver is initialized with `new Logic.Solver()`.
- The satisfiability check is performed synchronously using `solver.solve()`.
- If the formula is satisfiable, `solve()` returns a solution object, from which the array of true variables can be retrieved using `getTrueVars()`. Otherwise, `null` is returned.

As an example, the formula 
$$ (\neg p \vee  q \vee \neg r) \wedge (p \vee \neg q \vee r) \wedge (p \vee q \vee \neg r) \wedge (p \vee q \vee r), $$
which is in conjunctive normal form, is represented by the following code:

In [3]:
const solver = new Logic.Solver();
solver.require(Logic.or("-p", "q", "-r"));
solver.require(Logic.or("p", "-q", "r"));
solver.require(Logic.or("p", "q", "-r"));
solver.require(Logic.or("p", "q", "r"));

In [4]:
const sol = solver.solve();

In [5]:
sol.getTrueVars()

[ 'p' ]


This shows that the formula `f` is satisfiable and that the  propositional valuation 
$$ \mathcal{I} = \{ p \mapsto \texttt{True}, q \mapsto \texttt{False}, r \mapsto \texttt{False} \} $$
is a solution for `f`, i.e. we have
$$ \mathcal{I}(\texttt{f}) = \texttt{True}. $$

## Sudoku

In [6]:
import { RecursiveSet, Value, flatMap } from 'recursive-set';

type Variable        = string;
type NegatedVariable = `-${string}`; 
type Literal         = Variable | NegatedVariable;
type Clause          = RecursiveSet<Literal>;

In [7]:
type RS<T extends Value> = RecursiveSet<T>;

In [8]:
function empty<T extends Value>(): RS<T> {
    return new RecursiveSet<T>()
}

The finish mathematician Arto Inkala claims to have created the [hardest sudoku](https://abcnews.go.com/blogs/headlines/2012/06/can-you-solve-the-hardest-ever-sudoku) ever.  It is defined below.

In [9]:
function createPuzzle(): (number | string)[][] {
  return [[ 8 , "*", "*", "*", "*", "*", "*", "*", "*"], 
          ["*", "*",  3 ,  6 , "*", "*", "*", "*", "*"],
          ["*",  7 , "*", "*",  9 , "*",  2 , "*", "*"],
          ["*",  5 , "*", "*", "*",  7 , "*", "*", "*"],
          ["*", "*", "*", "*",  4 ,  5 ,  7 , "*", "*"],
          ["*", "*", "*",  1 , "*", "*", "*",  3 , "*"],
          ["*", "*",  1 , "*", "*", "*", "*",  6 ,  8 ],
          ["*", "*",  8 ,  5 , "*", "*", "*",  1 , "*"],
          ["*",  9 , "*", "*", "*", "*",  4 , "*", "*"]
         ];
}

We use the following variables:
* `Q<r,c,d>` is a Boolean variable stating that the field in row `r` and column `c` holds the digit `d`.
  Here, `r`, `c`, `d` are all elements from the set $\{1,\cdots,9\}$.
    
The function `varName(row, col, digit)` returns a formated string that is interpreted as a variable name.

In [10]:
function varName(row: number, col: number, digit: number): string {
    return `Q<${row},${col},${digit}>`;
}

In [11]:
varName(1,2,3);

Q<1,2,3>


In [12]:
function complement(l: Literal): Literal {
    if (l.startsWith('-')) {
        return l.substring(1);
    } else {
        return '-' + l;
    }
}

The function `atMostOne(S)` takes a set `S` of propositional variables as its argument.  It returns a set of clauses
expressing the fact that at most one of the variables of `S` is true.

In [13]:
function atMostOne(S: RS<Variable>): RS<Clause> {
    const result = empty<Clause>();
    for (const p of S) {
        for (const q of S) {
            if (p < q) {
                const lit1 = complement(p);
                const lit2 = complement(q);
                result.add(new RecursiveSet<Literal>(lit1, lit2));
            }
        }
    }
    return result;
}

The function `atLeastOne(S)` takes a set `S` of propositional variables as its argument.  It returns a set of clauses
expressing the fact that at least one of the variables of `S` is true.

In [14]:
function atLeastOne(S: RS<Variable>): RS<Clause> {
    return new RecursiveSet<Clause>(new RecursiveSet<Literal>(...S));
}

The function `exactlyOne(S)` takes a set `S` of propositional variables as its argument.  It returns a set of clauses
expressing the fact that exactly one of the variables of `S` is true.

In [15]:
function exactlyOne(S: RS<Variable>): RS<Clause> {
    const atMost  = atMostOne(S);
    const atLeast = atLeastOne(S);
    return atMost.union(atLeast);
}

In [16]:
exactlyOne(new RecursiveSet<Variable>('a', 'b', 'c'));

{{-a, -b}, {-a, -c}, {-b, -c}, {a, b, c}}


The function `exactlyOneForDigit` takes two arguments:
* An array `L` of pairs of indices representing Sudoku coordinates   as its first argument.
  The elements of `L` are pairs of the form `(row, col)`, where both `row` and `col` are elements of the set $\{1, \cdots, 9\}$.
* A specific `digit` as its second arguments.  
  `digit` is an element from the set `{1,...,9}`.

It returns a set of clauses expressing the fact that the given `digit` 
appears exactly once among the fields specified by the coordinate pairs in `L`.

In [17]:
function exactlyOneForDigit(L: Array<[number, number]>, digit: number): RS<Clause> {
    return exactlyOne(new RecursiveSet<Variable>(
        ...L.map(([row, col]) => varName(row, col, digit))
    ));
}

The function `exactlyOnce` takes an array `L` of pairs of indices as its argument.  The elements of `L` are pairs of the form
`(row, col)`, where both `row` and `col` are elements of the set $\{1, \cdots, 9\}$.
It returns a set of formulas expressing that all Sudoku fields specified by the coordinate pairs in `L` take different digits as values.

In [18]:
function exactlyOnce(L: Array<[number, number]>): RS<Clause> {
    const digits = [1, 2, 3, 4, 5, 6, 7, 8, 9];
    return digits.reduce(
        (clauses, digit) => clauses.union(exactlyOneForDigit(L, digit)),
        empty<Clause>()
    );
}

In [19]:
const result = exactlyOnce(
    Array.from({ length: 9 }, (_, i) => [1, i + 1] as [number, number])
);
for (const clause of result) {
    console.log(`{${[...clause].map(lit => JSON.stringify(lit)).join(', ')}}`);
}

{"-Q<1,1,1>", "-Q<1,2,1>"}
{"-Q<1,1,1>", "-Q<1,3,1>"}
{"-Q<1,1,1>", "-Q<1,4,1>"}
{"-Q<1,1,1>", "-Q<1,5,1>"}
{"-Q<1,1,1>", "-Q<1,6,1>"}
{"-Q<1,1,1>", "-Q<1,7,1>"}
{"-Q<1,1,1>", "-Q<1,8,1>"}
{"-Q<1,1,1>", "-Q<1,9,1>"}
{"-Q<1,2,1>", "-Q<1,3,1>"}
{"-Q<1,2,1>", "-Q<1,4,1>"}
{"-Q<1,2,1>", "-Q<1,5,1>"}
{"-Q<1,2,1>", "-Q<1,6,1>"}
{"-Q<1,2,1>", "-Q<1,7,1>"}
{"-Q<1,2,1>", "-Q<1,8,1>"}
{"-Q<1,2,1>", "-Q<1,9,1>"}
{"-Q<1,3,1>", "-Q<1,4,1>"}
{"-Q<1,3,1>", "-Q<1,5,1>"}
{"-Q<1,3,1>", "-Q<1,6,1>"}
{"-Q<1,3,1>", "-Q<1,7,1>"}
{"-Q<1,3,1>", "-Q<1,8,1>"}
{"-Q<1,3,1>", "-Q<1,9,1>"}
{"-Q<1,4,1>", "-Q<1,5,1>"}
{"-Q<1,4,1>", "-Q<1,6,1>"}
{"-Q<1,4,1>", "-Q<1,7,1>"}
{"-Q<1,4,1>", "-Q<1,8,1>"}
{"-Q<1,4,1>", "-Q<1,9,1>"}
{"-Q<1,5,1>", "-Q<1,6,1>"}
{"-Q<1,5,1>", "-Q<1,7,1>"}
{"-Q<1,5,1>", "-Q<1,8,1>"}
{"-Q<1,5,1>", "-Q<1,9,1>"}
{"-Q<1,6,1>", "-Q<1,7,1>"}
{"-Q<1,6,1>", "-Q<1,8,1>"}
{"-Q<1,6,1>", "-Q<1,9,1>"}
{"-Q<1,7,1>", "-Q<1,8,1>"}
{"-Q<1,7,1>", "-Q<1,9,1>"}
{"-Q<1,8,1>", "-Q<1,9,1>"}
{"Q<1,1,1>", "Q<1,2,1>", "Q<

{"-Q<1,4,8>", "-Q<1,8,8>"}
{"-Q<1,4,8>", "-Q<1,9,8>"}
{"-Q<1,5,8>", "-Q<1,6,8>"}
{"-Q<1,5,8>", "-Q<1,7,8>"}
{"-Q<1,5,8>", "-Q<1,8,8>"}
{"-Q<1,5,8>", "-Q<1,9,8>"}
{"-Q<1,6,8>", "-Q<1,7,8>"}
{"-Q<1,6,8>", "-Q<1,8,8>"}
{"-Q<1,6,8>", "-Q<1,9,8>"}
{"-Q<1,7,8>", "-Q<1,8,8>"}
{"-Q<1,7,8>", "-Q<1,9,8>"}
{"-Q<1,8,8>", "-Q<1,9,8>"}
{"Q<1,1,8>", "Q<1,2,8>", "Q<1,3,8>", "Q<1,4,8>", "Q<1,5,8>", "Q<1,6,8>", "Q<1,7,8>", "Q<1,8,8>", "Q<1,9,8>"}
{"-Q<1,1,9>", "-Q<1,2,9>"}
{"-Q<1,1,9>", "-Q<1,3,9>"}
{"-Q<1,1,9>", "-Q<1,4,9>"}
{"-Q<1,1,9>", "-Q<1,5,9>"}
{"-Q<1,1,9>", "-Q<1,6,9>"}
{"-Q<1,1,9>", "-Q<1,7,9>"}
{"-Q<1,1,9>", "-Q<1,8,9>"}
{"-Q<1,1,9>", "-Q<1,9,9>"}
{"-Q<1,2,9>", "-Q<1,3,9>"}
{"-Q<1,2,9>", "-Q<1,4,9>"}
{"-Q<1,2,9>", "-Q<1,5,9>"}
{"-Q<1,2,9>", "-Q<1,6,9>"}
{"-Q<1,2,9>", "-Q<1,7,9>"}
{"-Q<1,2,9>", "-Q<1,8,9>"}
{"-Q<1,2,9>", "-Q<1,9,9>"}
{"-Q<1,3,9>", "-Q<1,4,9>"}
{"-Q<1,3,9>", "-Q<1,5,9>"}
{"-Q<1,3,9>", "-Q<1,6,9>"}
{"-Q<1,3,9>", "-Q<1,7,9>"}
{"-Q<1,3,9>", "-Q<1,8,9>"}
{"-Q<1,3,9>", "-Q<1,9,9>"}


The function `exactlyOneDigit(row, col)` takes integers `row` and `col` as arguments.  These specify the row and column of a field in a Sudoku.  The function returns a set of clauses specifying that exactly one of the variables

* `Q<row,col,1>`, `Q<row,col,2>`, $\cdots$, `Q<row,col,9>`

is `True`.

In [20]:
function exactlyOneDigit(row: number, col: number): RecursiveSet<Clause> {
    const digits = [1, 2, 3, 4, 5, 6, 7, 8, 9];
    return exactlyOne(new RecursiveSet<Variable>(
        ...digits.map(digit => varName(row, col, digit))));
}

In [21]:
for (const clause of exactlyOneDigit(1, 1)) {
    console.log(clause.toString()); 
}

{-Q<1,1,1>, -Q<1,1,2>}
{-Q<1,1,1>, -Q<1,1,3>}
{-Q<1,1,1>, -Q<1,1,4>}
{-Q<1,1,1>, -Q<1,1,5>}
{-Q<1,1,1>, -Q<1,1,6>}
{-Q<1,1,1>, -Q<1,1,7>}
{-Q<1,1,1>, -Q<1,1,8>}
{-Q<1,1,1>, -Q<1,1,9>}
{-Q<1,1,2>, -Q<1,1,3>}
{-Q<1,1,2>, -Q<1,1,4>}
{-Q<1,1,2>, -Q<1,1,5>}
{-Q<1,1,2>, -Q<1,1,6>}
{-Q<1,1,2>, -Q<1,1,7>}
{-Q<1,1,2>, -Q<1,1,8>}
{-Q<1,1,2>, -Q<1,1,9>}
{-Q<1,1,3>, -Q<1,1,4>}
{-Q<1,1,3>, -Q<1,1,5>}
{-Q<1,1,3>, -Q<1,1,6>}
{-Q<1,1,3>, -Q<1,1,7>}
{-Q<1,1,3>, -Q<1,1,8>}
{-Q<1,1,3>, -Q<1,1,9>}
{-Q<1,1,4>, -Q<1,1,5>}
{-Q<1,1,4>, -Q<1,1,6>}
{-Q<1,1,4>, -Q<1,1,7>}
{-Q<1,1,4>, -Q<1,1,8>}
{-Q<1,1,4>, -Q<1,1,9>}
{-Q<1,1,5>, -Q<1,1,6>}
{-Q<1,1,5>, -Q<1,1,7>}
{-Q<1,1,5>, -Q<1,1,8>}
{-Q<1,1,5>, -Q<1,1,9>}
{-Q<1,1,6>, -Q<1,1,7>}
{-Q<1,1,6>, -Q<1,1,8>}
{-Q<1,1,6>, -Q<1,1,9>}
{-Q<1,1,7>, -Q<1,1,8>}
{-Q<1,1,7>, -Q<1,1,9>}
{-Q<1,1,8>, -Q<1,1,9>}
{Q<1,1,1>, Q<1,1,2>, Q<1,1,3>, Q<1,1,4>, Q<1,1,5>, Q<1,1,6>, Q<1,1,7>, Q<1,1,8>, Q<1,1,9>}


The function `constraintsFromPuzzle`  returns a set of clauses stating that the variables corresponding to numbers that are already given in the Sudoku puzzle take the values that are specified.

In [22]:
function constraintsFromPuzzle(): RS<Clause> {
    const Puzzle = createPuzzle();
    const Clauses = empty<Clause>();
    for (let row = 0; row < 9; row++) {
        for (let col = 0; col < 9; col++) {
            const value = Puzzle[row][col];
            if (value != '*') {
                const v = varName(row + 1, col + 1, value as number);
                const unitClause = new RecursiveSet<Literal>(v);
                Clauses.add(unitClause);
            }
        }
    }
    return Clauses;
}

In [23]:
for (const clause of constraintsFromPuzzle()) {
    console.log(clause.toString()); 
}

{Q<1,1,8>}
{Q<2,3,3>}
{Q<2,4,6>}
{Q<3,2,7>}
{Q<3,5,9>}
{Q<3,7,2>}
{Q<4,2,5>}
{Q<4,6,7>}
{Q<5,5,4>}
{Q<5,6,5>}
{Q<5,7,7>}
{Q<6,4,1>}
{Q<6,8,3>}
{Q<7,3,1>}
{Q<7,8,6>}
{Q<7,9,8>}
{Q<8,3,8>}
{Q<8,4,5>}
{Q<8,8,1>}
{Q<9,2,9>}
{Q<9,7,4>}


In [24]:
const L = [1, 2, 3, 4, 5, 6, 7, 8, 9]; 
L.flatMap(r => L.map(c => [r, c]));

[
  [ 1, 1 ], [ 1, 2 ], [ 1, 3 ], [ 1, 4 ], [ 1, 5 ],
  [ 1, 6 ], [ 1, 7 ], [ 1, 8 ], [ 1, 9 ], [ 2, 1 ],
  [ 2, 2 ], [ 2, 3 ], [ 2, 4 ], [ 2, 5 ], [ 2, 6 ],
  [ 2, 7 ], [ 2, 8 ], [ 2, 9 ], [ 3, 1 ], [ 3, 2 ],
  [ 3, 3 ], [ 3, 4 ], [ 3, 5 ], [ 3, 6 ], [ 3, 7 ],
  [ 3, 8 ], [ 3, 9 ], [ 4, 1 ], [ 4, 2 ], [ 4, 3 ],
  [ 4, 4 ], [ 4, 5 ], [ 4, 6 ], [ 4, 7 ], [ 4, 8 ],
  [ 4, 9 ], [ 5, 1 ], [ 5, 2 ], [ 5, 3 ], [ 5, 4 ],
  [ 5, 5 ], [ 5, 6 ], [ 5, 7 ], [ 5, 8 ], [ 5, 9 ],
  [ 6, 1 ], [ 6, 2 ], [ 6, 3 ], [ 6, 4 ], [ 6, 5 ],
  [ 6, 6 ], [ 6, 7 ], [ 6, 8 ], [ 6, 9 ], [ 7, 1 ],
  [ 7, 2 ], [ 7, 3 ], [ 7, 4 ], [ 7, 5 ], [ 7, 6 ],
  [ 7, 7 ], [ 7, 8 ], [ 7, 9 ], [ 8, 1 ], [ 8, 2 ],
  [ 8, 3 ], [ 8, 4 ], [ 8, 5 ], [ 8, 6 ], [ 8, 7 ],
  [ 8, 8 ], [ 8, 9 ], [ 9, 1 ], [ 9, 2 ], [ 9, 3 ],
  [ 9, 4 ], [ 9, 5 ], [ 9, 6 ], [ 9, 7 ], [ 9, 8 ],
  [ 9, 9 ]
]


The functiuon `getBlockCells` calculates the absolute grid coordinates for all 9 cells 
within a specific 3x3 Sudoku block.  The Sudoku board is divided into a 3x3 grid of blocks. 
This function takes the zero-based indices of a block and maps its local cells to the 
1-based coordinates used by the solver variables. 
 * `r` - The row index of the 3x3 macro-block (must be 0, 1, or 2).
 * `c` - The column index of the 3x3 macro-block (must be 0, 1, or 2).
The function returns an array of 9 tuples, where each tuple `[row, col]` represents 
 the 1-based row and column coordinates on the full 9x9 Sudoku board.
 
 Example
 Get the coordinates for the top-middle block
 ```getBlockCells(0, 1) = 
         [[1, 4], [1, 5], [1, 6], 
          [2, 4], [2, 5], [2, 6], 
          [3, 4], [3, 5], [3, 6]]
```

In [25]:
function getBlockCells(r: number, c: number): Array<[number, number]> {
    const baseIndices = new RecursiveSet<number>(1, 2, 3);
    const basePairs   = baseIndices.cartesianProduct(baseIndices);
    return [...basePairs].map(
        ([row, col]) => [r * 3 + row, c * 3 + col]
    );
}

In [26]:
getBlockCells(0, 0);

[
  [ 1, 1 ], [ 1, 2 ],
  [ 1, 3 ], [ 2, 1 ],
  [ 2, 2 ], [ 2, 3 ],
  [ 3, 1 ], [ 3, 2 ],
  [ 3, 3 ]
]


The function `allConstraints` returns a CSP that encodes the given sudoku as a CSP.

In [27]:
function allConstraints(): RS<Clause> {
    const L = [1, 2, 3, 4, 5, 6, 7, 8, 9];
    // 1. Start with constraints from the puzzle
    let Clauses = constraintsFromPuzzle();
    // 2. There is exactly one digit in every field
    const allCoords = L.flatMap(r => L.map(c => [r, c]));
    Clauses = flatMap(allCoords, ([row, col]) => exactlyOneDigit(row, col), Clauses);
    // 3. All entries in a row are unique
    Clauses = flatMap(L, row => exactlyOnce(L.map(col => [row, col])), Clauses);
    // 4. All entries in a column are unique
    Clauses = flatMap(L, col => exactlyOnce(L.map(row => [row, col])), Clauses);
    // 5. All entries in a 3x3 square are unique
    const outerIndices     = new RecursiveSet<number>(0, 1, 2);
    const blockCoordinates = outerIndices.cartesianProduct(outerIndices);
    Clauses = flatMap(blockCoordinates, ([r, c]) => exactlyOnce(getBlockCells(r, c)), Clauses);
    return Clauses;
}

In [28]:
const clauses = allConstraints();
console.log("--- Clauses with size 1 ---");
for (const clause of clauses) {
    if (clause.size === 1) {
        console.log(clause.toString());
    }
}
console.log("\n--- Clauses with size 9 ---");
for (const clause of clauses) {
    if (clause.size === 9) {
        console.log(clause.toString());
    }
}

--- Clauses with size 1 ---
{Q<1,1,8>}
{Q<2,3,3>}
{Q<2,4,6>}
{Q<3,2,7>}
{Q<3,5,9>}
{Q<3,7,2>}
{Q<4,2,5>}
{Q<4,6,7>}
{Q<5,5,4>}
{Q<5,6,5>}
{Q<5,7,7>}
{Q<6,4,1>}
{Q<6,8,3>}
{Q<7,3,1>}
{Q<7,8,6>}
{Q<7,9,8>}
{Q<8,3,8>}
{Q<8,4,5>}
{Q<8,8,1>}
{Q<9,2,9>}
{Q<9,7,4>}

--- Clauses with size 9 ---
{Q<1,1,1>, Q<1,1,2>, Q<1,1,3>, Q<1,1,4>, Q<1,1,5>, Q<1,1,6>, Q<1,1,7>, Q<1,1,8>, Q<1,1,9>}
{Q<1,2,1>, Q<1,2,2>, Q<1,2,3>, Q<1,2,4>, Q<1,2,5>, Q<1,2,6>, Q<1,2,7>, Q<1,2,8>, Q<1,2,9>}
{Q<1,3,1>, Q<1,3,2>, Q<1,3,3>, Q<1,3,4>, Q<1,3,5>, Q<1,3,6>, Q<1,3,7>, Q<1,3,8>, Q<1,3,9>}
{Q<1,4,1>, Q<1,4,2>, Q<1,4,3>, Q<1,4,4>, Q<1,4,5>, Q<1,4,6>, Q<1,4,7>, Q<1,4,8>, Q<1,4,9>}
{Q<1,5,1>, Q<1,5,2>, Q<1,5,3>, Q<1,5,4>, Q<1,5,5>, Q<1,5,6>, Q<1,5,7>, Q<1,5,8>, Q<1,5,9>}
{Q<1,6,1>, Q<1,6,2>, Q<1,6,3>, Q<1,6,4>, Q<1,6,5>, Q<1,6,6>, Q<1,6,7>, Q<1,6,8>, Q<1,6,9>}
{Q<1,7,1>, Q<1,7,2>, Q<1,7,3>, Q<1,7,4>, Q<1,7,5>, Q<1,7,6>, Q<1,7,7>, Q<1,7,8>, Q<1,7,9>}
{Q<1,8,1>, Q<1,8,2>, Q<1,8,3>, Q<1,8,4>, Q<1,8,5>, Q<1,8,6>, Q<1,8,7>, Q<1,

In [29]:
clauses.size

10551


In the code below the clauses are iteratively processed to construct the logical formula for ``logic-solver``. For each clause, the terms are collected, handling negations explicitly by converting negative literals (e.g., `¬Q<1,1,1>`) into `-Q<1,1,1>`.

In [30]:
const solver = new Logic.Solver();
for (const clause of clauses) {
    const terms: (string | Logic.Formula)[] = [];
    for (const literal of clause) {
        if (literal.startsWith('¬')) {
            const varName = literal.substring(1);
            terms.push("-" + varName);
        } else {
            terms.push(literal);
        }
    }
    solver.require(Logic.or(...terms));
}

In [31]:
9**3

729


Even though this Sudoku is modelled using $9^3 = 729$ propositional variables and we have 10551 clauses, logic-solver uses about 15 milliseconds to solve the problem. 

In [32]:
console.time('Solving');
const sol = solver.solve();
console.timeEnd('Solving');
if (sol === null) {
    throw new Error("No solution found!");
}
const Solution = sol.getTrueVars();

Solving: 23.883ms


In [33]:
Solution

[
  'Q<1,1,8>', 'Q<1,2,1>', 'Q<1,3,2>', 'Q<1,4,7>', 'Q<1,5,5>',
  'Q<1,6,3>', 'Q<1,7,6>', 'Q<1,8,4>', 'Q<1,9,9>', 'Q<2,1,9>',
  'Q<2,2,4>', 'Q<2,3,3>', 'Q<2,4,6>', 'Q<2,5,8>', 'Q<2,6,2>',
  'Q<2,7,1>', 'Q<2,8,7>', 'Q<2,9,5>', 'Q<3,1,6>', 'Q<3,2,7>',
  'Q<3,3,5>', 'Q<3,4,4>', 'Q<3,5,9>', 'Q<3,6,1>', 'Q<3,7,2>',
  'Q<3,8,8>', 'Q<3,9,3>', 'Q<4,1,1>', 'Q<4,2,5>', 'Q<4,3,4>',
  'Q<4,4,2>', 'Q<4,5,3>', 'Q<4,6,7>', 'Q<4,7,8>', 'Q<4,8,9>',
  'Q<4,9,6>', 'Q<5,1,3>', 'Q<5,2,6>', 'Q<5,3,9>', 'Q<5,4,8>',
  'Q<5,5,4>', 'Q<5,6,5>', 'Q<5,7,7>', 'Q<5,8,2>', 'Q<5,9,1>',
  'Q<6,1,2>', 'Q<6,2,8>', 'Q<6,3,7>', 'Q<6,4,1>', 'Q<6,5,6>',
  'Q<6,6,9>', 'Q<6,7,5>', 'Q<6,8,3>', 'Q<6,9,4>', 'Q<7,1,5>',
  'Q<7,2,2>', 'Q<7,3,1>', 'Q<7,4,9>', 'Q<7,5,7>', 'Q<7,6,4>',
  'Q<7,7,3>', 'Q<7,8,6>', 'Q<7,9,8>', 'Q<8,1,4>', 'Q<8,2,3>',
  'Q<8,3,8>', 'Q<8,4,5>', 'Q<8,5,2>', 'Q<8,6,6>', 'Q<8,7,9>',
  'Q<8,8,1>', 'Q<8,9,7>', 'Q<9,1,7>', 'Q<9,2,9>', 'Q<9,3,6>',
  'Q<9,4,3>', 'Q<9,5,1>', 'Q<9,6,8>', 'Q<9,7,4>', 'Q<9,8,5>',
  'Q<9

## Graphical Representation

In [34]:
import { display } from 'tslab';
function showSolution(Solution: string[], width: string = '50%'): void {
    const Sudoku = createPuzzle();
    const solutionMap: Record<string, number> = {};
    for (const literal of Solution) {
        const match = literal.match(/^Q<(\d+),(\d+),(\d+)>$/);
        if (match) {
            const row = parseInt(match[1], 10);
            const col = parseInt(match[2], 10);
            const digit = parseInt(match[3], 10);
            solutionMap[`V${row}${col}`] = digit;
        }
    }
    let html = `<table style="width:${width}; border-collapse: collapse; border: 2px solid black; font-family: sans-serif;">`;
    for (let row = 0; row < 9; row++) {
        html += '<tr>';
        for (let col = 0; col < 9; col++) {
            const key = `V${row + 1}${col + 1}`;
            let value = solutionMap[key];
            const original = Sudoku[row][col];
            let cellStyle = "font-weight: normal;";
            if (original !== '*') {
                value = original as number;
                cellStyle = "font-weight: bold;";
            }
            const blockRow = Math.floor(row / 3);
            const blockCol = Math.floor(col / 3);
            const isGray = (blockRow + blockCol) % 2 !== 0;
            const bgColor = isGray ? '#f0f0f0' : '#ffffff';
            let borderStyle = "border: 1px solid #ccc;";
            if ((col + 1) % 3 === 0 && col < 8) borderStyle += "border-right: 2px solid black;";
            if ((row + 1) % 3 === 0 && row < 8) borderStyle += "border-bottom: 2px solid black;";
            html += `<td style="${borderStyle} width:30px; height:30px; text-align:center; font-size:20px; background-color:${bgColor}; ${cellStyle}">${value || ''}</td>`;
        }
        html += '</tr>';
    }
    html += '</table>';
    display.html(html);
}

In [35]:
showSolution(Solution, "50%");

8,1,2,7,5,3,6,4,9
9,4,3,6,8,2,1,7,5
6,7,5,4,9,1,2,8,3
1,5,4,2,3,7,8,9,6
3,6,9,8,4,5,7,2,1
2,8,7,1,6,9,5,3,4
5,2,1,9,7,4,3,6,8
4,3,8,5,2,6,9,1,7
7,9,6,3,1,8,4,5,2
